loadbalancer/algorithm.go

```diff
@@ -168,32 +72,30 @@ func (r *roundRobin) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 		return ctx.Route.LBEndpoints[0]
 	}
 
+	ctx.Route.Lock()
+	defer ctx.Route.Unlock()
+
 	r.mx.Lock()
 	defer r.mx.Unlock()
-	r.index = (r.index + 1) % len(ctx.Route.LBEndpoints)
 
-	if ctx.Route.LBFadeInDuration <= 0 {
-		return ctx.Route.LBEndpoints[r.index]
+	endpoints := ctx.Route.GoodEndpoints
```

Comment:
With this refactoring the list of endpoints is variable between calls to Apply.
We need to make sure that algorithms continue to behave as expected (esp. consitent hash that uses pre-calculated datastructure based on the full list of endpoints).